corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+Elasticsearch indexes in a way that is transparent to parts of HQ that write to
+Elasticsearch (e.g. pillowtop).  Reindexing is necessary for making changes to
+index mappings, is a prerequisite to upgrading an Elasticsearch cluster, and is
+also needed for changing low-level index configurations (e.g. sharding).
+
+There is an existing procedure draft that documents the steps that were used on
+one occasion to reindex the ``case_search`` index.  This procedure leveraged a
+custom pillow to "backfill" the cloned index (i.e. initially populated using
+Elasticsearch Reindex API). That procedure only works for a subset of HQ
+Elasticsearch indexes, and is too risky to be considered as an ongoing Elastic
+maintenance strategy. There are several key constraints that an HQ reindexing
+procedure should meet which the existing procedure does not:
+
+- simple and robust
+- performed with standard maintenance practices
+- provides the ability to test and verify the integrity of a new index before it
+  is too late to be rejected
+- allows HQ Elasticsearch index state to remain decoupled from the
+  commcare-cloud codebase
+- is not disruptive -- does not prohibit any other kind of standard maintenance
+  that might come up while the operation is underway
+- is "fire and forget" -- does not require active polling of intermediate state
+  in order to progress the overall operation
+- is practical for third party HQ hosters to use
+
+One way to accomplish these constraints is to implement an "index multiplexing"
+feature in HQ, where Elasticsearch write operations are duplicated across two
+indexes. This design facilitates maintaining two up-to-date versions of any
+index (a primary read/write index and a secondary write-only index),
+allowing HQ to run in a "normal" state (i.e. not a custom "maintenance" state)
+while providing the ability to switch back and forth (swapping primary and
+secondary) before fully committing to abandoning one of them. Creating a copy of
+an index is the unavoidable nature of a reindex operation, and multiplexing
+allows safe switching from one to the other without causing disruptions or
+outages while keeping both up-to-date.
+
+The least disruptive way to accomplish a multiplexing design is with an adapter
+layer that operates between the low-level third party Elasticsearch Python
+client library and high-level HQ components which need to read/write data in an
+Elasticsearch index. Conveniently, HQ already has the initial framework for this
+layer (the ``ElasticsearchInterface`` class), so the adapter layer is not a new
+concept. The reason the ``ElasticsearchInterface`` implementation cannot
+accommodate multiplexing is because it is not widely adopted for low-level
+Elasticsearch operations (e.g. index creation and configuration), and it does
+not sufficiently decouple low-level Elasticsearch document manipulation logic
+from the high-level HQ code that uses it.
```

Comment:
Many areas of Elastic-related code in HQ simply step around the `ElasticsearchInterface` object by using the low-level `elasticsearch-py` client object directly. Multiplexing requires that the intermediate adapter absolutely _know_ everything that happens to documents in an index (so it can intelligently replicate that action in a secondary index).

Reply:
In understand that, my question is why create a new construct instead of make sure the existing one is used everywhere.

Reply:
Because the existing `ElasticsearchInterface` implementation is the wrong level of abstraction for multiplexing (it's too low level for this functionality).